### PR TITLE
evaluate required_consts when pushing stack frame in Miri engine

### PR DESCRIPTION
**Pull request by RalfJung**
[Just like codegen](https://github.com/rust-lang/rust/pull/70820/files#diff-32c57af5c8e23eb048f55d1e955e5cd5R194), Miri needs to make sure all `required_consts` evaluate successfully, to catch post-monomorphization errors.

While at it I also moved the const_eval error reporting logic into rustc_mir::const_eval::error; there is no reason it should be in `rustc_middle`. I kept this in a separate commit for easier reviewing.

Helps with https://github.com/rust-lang/miri/issues/1382. I will add a test on the Miri side (done now: https://github.com/rust-lang/miri/pull/1504).
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/eval_context.rs [hunk 1]
```diff
@@ -625,6 +652,13 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         let frame = M::init_frame_extra(self, pre_frame)?;
         self.stack_mut().push(frame);
 
+        // Make sure all the constants required by this frame evaluate successfully (post-monomorphization check).
+        for const_ in &body.required_consts {
+            let const_ =
+                self.subst_from_current_frame_and_normalize_erasing_regions(const_.literal);
+            self.const_to_op(const_, None)?;
+        }
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs:660]: This is the key new loop.

**Comment by oli-obk** [comment]
```rust
#![warn(const_err, unconditional_panic)]

struct PrintName<T>(T);
impl<T> PrintName<T> {
    const VOID: () = [()][2]; //~WARN any use of this value will cause an error
}

const fn no_codegen<T>() {
    if false {
        let _ = PrintName::<T>::VOID; //~ERROR referenced constant has errors
    }
}

pub static FOO: () = no_codegen::<i32>();

fn main() {
    FOO
}
```

works on beta, so we should probably beta backport this (and add the above as a test).

The beta backport should not contain the first commit though

**Comment by RalfJung** [comment]
I don't see why this should be beta-backported -- it's an ooold bug for that to work.

**Comment by oli-obk** [comment]
>  it's an ooold bug for that to work.

not really, before this beta there was no conditional execution. So the constant was always evaluated unless optimized out... ok... true, we used to optimize them out. So nevermind on the beta nomination, but the test should still be in this repo, not just in miri

**Comment by oli-obk** [comment]
So this is a breaking change, but only if your code was UB anyway

**Comment by RalfJung** [comment]
> So this is a breaking change, but only if your code was UB anyway

It is, just like the corresponding change in codegen was a breaking change.
I am not sure if the old code was UB (dead code cannot cause UB), but it is a bug that we fixed for codegen and should likewise fix for CTFE.

> the test should still be in this repo, not just in miri

Fair, I'll add it.

**Comment by RalfJung** [comment]
@oli-obk I added a test.

**Pull request merged**